Auth to upload guide (#97)
AdobeDocs/frameio-api: src/pages/guides/Authorization to File Upload/index.md

![Request](../image_15.png)

![Request 2](../image_16.png)

## Multi-part Upload  

After creating a placeholder file through the **Create File (local upload)** endpoint, you will have one or more upload urls. These upload urls are required to make upload requests. It may be useful to compose a shell script that splits up the source file into chunks and issues the same number of subsequent requests.

> **NOTE:** These are important details to keep in mind when sending the subsequent upload requests.

> * The HTTP request method must be `PUT`.
> * The `x-amz-acl` header must be included and be set to private.
> * The `Content-Type` header must match the `media_type` specified in the original **Create File (local upload)** request. This is true even when uploading the file as separate parts.

In the sample Python script below, we're passing in one upload url in the `upload_urls` parameter, though as noted above, you may have more than one provided to you based on the file size set in the request body for **Create File (local upload).**

``` python
import requests
import math
from typing import List
from tqdm import tqdm  # For progress bar

> def upload_file_in_chunks(file_path, upload_urls, chunk_size) -> bool:
>     """
>     Upload a file in chunks using presigned URLs.
>     """
>     try:
>         # Get file size
>         with open(file_path, 'rb') as f:
>             f.seek(0, 2)  # Seek to end of file
>             file_size = f.tell()
>         
>         # Calculate chunk size if not provided
>         if chunk_size is None:
>             chunk_size = math.ceil(file_size / len(upload_urls))
>         
>         print(f"File size: {file_size} bytes")
>         print(f"Chunk size: {chunk_size} bytes")
>         print(f"Number of chunks: {len(upload_urls)}")
>         
>         # Upload each chunk
>         with open(file_path, 'rb') as f:
>             with tqdm(total=len(upload_urls), desc="Uploading chunks") as pbar:
>                 for i, url in enumerate(upload_urls):
>                     start_byte = i * chunk_size
>                     end_byte = min(start_byte + chunk_size, file_size)
>                     
>                     # Read chunk from file
>                     f.seek(start_byte)
>                     chunk = f.read(end_byte - start_byte)
>                     
>                     # Upload chunk
>                     response = requests.put(
>                         url,
>                         data=chunk,
>                         headers={
>                             'Content-Length': str(len(chunk)),
>                             'Content-Type': 'image/jpeg',
>                             'X-Amz-Acl': 'private'
>                             }
>                     )
>                     
>                     if response.status_code != 200:
>                         print(f"Failed to upload chunk {i+1}. Status code: {response.status_code}")
>                         return False
>                     
>                     pbar.update(1)
>         
>         return True
>     
>     except Exception as e:
>         print(f"Error during upload: {str(e)}")
>         return False

Reviewer: ```suggestion
def upload_file_in_chunks(file_path: str, upload_urls: list[str], content_type: str | None = None, chunk_size: int | None = None) -> bool:
    """
    Upload a file in chunks using presigned URLs.
    """
    try:
        # Auto-detect content type based on file extension
        if content_type is None:
            detected_content_type, _ = mimetypes.guess_type(file_path)
            content_type = detected_content_type # Default fallback
        
        print(f"Detected content type: {content_type}")
        
        # Get file size
        with open(file_path, 'rb') as f:
            f.seek(0, 2)  # Seek to end of file
            file_size = f.tell()
        
        # Calculate chunk size if not provided
        if chunk_size is None:
            chunk_size = math.ceil(file_size / len(upload_urls))
        
        print(f"File size: {file_size} bytes")
        print(f"Chunk size: {chunk_size} bytes")
        print(f"Number of chunks: {len(upload_urls)}")
        
        # Upload each chunk
        with open(file_path, 'rb') as f:
            with tqdm(total=len(upload_urls), desc="Uploading chunks") as pbar:
                for i, url in enumerate(upload_urls):
                    start_byte = i * chunk_size
                    end_byte = min(start_byte + chunk_size, file_size)
                    
                    # Read chunk from file
                    f.seek(start_byte)
                    chunk = f.read(end_byte - start_byte)
                    
                    print(f"Uploading chunk {i+1}: {len(chunk)} bytes")
                    
                    # Upload chunk with minimal headers matching the signature
                    response = requests.put(
                        url,
                        data=chunk,
                        headers={
                            'content-type': content_type,  
                            'x-amz-acl': 'private'       
                        }
                    )
                    
                    if response.status_code != 200:
                        print(f"Failed to upload chunk {i+1}. Status code: {response.status_code}")
                        print(f"Response text: {response.text}")
                        print(f"Response headers: {dict(response.headers)}")
                        return False
                    else:
                        print(f"Chunk {i+1} uploaded successfully!")
                    
                    pbar.update(1)
        
        return True
    
    except Exception as e:
        print(f"Error during upload: {str(e)}")
        return False
```

# Example usage
if __name__ == "__main__":
    # Replace these with your actual values
    file_path =  "/Users/MyComputer/local_upload/sample.jpg"  # Path to your file
    upload_urls = ["https://frameio-uploads-development.s3-accelerate.amazonaws.com/uploads/10e35b09-7e07-43a1-bd43-897f252ab627/original.png?..."]
    # Optional: specify chunk size in bytes (e.g., 10MB = 10 * 1024 * 1024)

>     chunk_size = 10 * 1024 * 1024  # 10MB

Reviewer: ```suggestion
    file_type = "image/jpeg"
    chunk_size = 10 * 1024 * 1024  # 10MB
```

# Authorization to File Upload Guide

This guide describes the authorization to file upload flow for the Frame.io V4 API.  

## Prerequisites

* [Overview of Adobe Developer Console](https://developer.adobe.com/developer-console/docs/guides/)
* [How To Refresh Access Tokens](https://developer.adobe.com/developer-console/docs/guides/authentication/UserAuthentication/ims#refreshing-access-tokens)
* [How To Upload a New File](https://developer.adobe.com/developer-console/docs/guides/)

> * [Fetching Access tokens](https://developer.adobe.com/developer-console/docs/guides/authentication/UserAuthentication/ims#fetching-access-tokens)

Reviewer 2: Shouldn't we place the guide for "Fetching Access tokens" before "How To Refresh Access Tokens"?

Author: Done.

> * [Uploading a New File](https://developer.adobe.com/frameio/guides/#uploading-a-new-file)

Author: Took out the **How to upload a new file** link.

## Adobe Authorization

Follow the guidance on the Adobe Developer Console [Authentication Guide](https://developer.adobe.com/developer-console/docs/guides/authentication/) to join and build with the Frame.io V4 API safely. Follow the steps in [Getting Started with the Adobe Developer Console](https://developer.adobe.com/developer-console/docs/guides/getting-started/) to create an Adobe Developer Console Project with a Frame.io API subscription configured with the appropriate credential type for your application.

## Credential and Token creation

> View, add, and manage credentials required for your application to connect safely and securely to external services. Configure your event registration to access authentication credentials. This could be a OAuth Server-to-Server credential or OAuth User authentication credential, depending on the type of event being configured.

Reviewer: Should there be a link here or anything?

Author: I link to all the things in the prereq section at the top. Does that work?

Author: The link is included in the prereqs at the top. That work @joannasese?

## Add an API

Using APIs allows your application to make calls to Adobe services by means of a REST API. APIs can be connected to your app in multiple ways, depending on the API or the type of app that you are building. To learn more about each of these authentication methods or API connections, read the [authentication documentation](https://developer.adobe.com/developer-console/docs/guides/authentication/).

* [Add an API using OAuth Server-to-Server credential](https://developer.adobe.com/developer-console/docs/guides/services/services-add-api-oauth-s2s)
* [Add an API using OAuth User authentication credential](https://developer.adobe.com/developer-console/docs/guides/services/services-add-api-oauth-user-authentication)
* [Add an API using API Key authentication](https://developer.adobe.com/developer-console/docs/guides/services/services-add-api-key)

## Upload a New File

There are two ways to upload a new file: `Create File (local upload)` and `Create File (remote upload)`.

**Local Upload**

To create a file through local upload, select the **Create File (local upload)** endpoint. The request body requires the file name and its file size.

> ![Request body](../image_12.png)

Reviewer 3: can this be a code snippet and not a screenshot?

Author: Updated 

![Screenshot 2025-06-27 at 07 46 21](https://github.com/user-attachments/assets/39471ff1-0e9c-446e-bf02-3d545d42c60e)

Reviewer 4: I've pulled the latest branch but am seeing this still as a png not a json code snippit, will address with my changes

* If the request is successful, a placeholder file resource is created without any content. Depending on the file size, the response body will include one or more `upload_urls`. See [Multi-part Upload](#multi-part-upload) for next steps.

> ![Upload url](../image_13.png)

Reviewer 3: Same with this, code snippets and not just a screenshot of code

Author: Updated
![Screenshot 2025-06-27 at 07 48 03](https://github.com/user-attachments/assets/0e1e96d9-a501-4075-94c0-df742072677d)

Reviewer 4: resolved in last commit

**Remote Upload**

To create a file through remote upload, select the **Create File (remote upload)** endpoint. The request body requires the file name and its source url.

> ![Remote upload](../image_14.png)

Reviewer: ```
{ 
    "data": {
        "name": "my_file.jpg",
        "source_url": "https://upload.wikimedia.org/wikipedia/commons/e/e1/White_Pixel_1x1.jpg"
    }
}
```

> ![Remote upload](../image_14.png)

Author: ![Screenshot 2025-06-27 at 07 49 47](https://github.com/user-attachments/assets/d7803a05-3a21-4787-9a74-ca0cd28dac2c)

* If the request is successful, the file appears in the UI.

> ![Request](../image_15.png)

Author: ![Screenshot 2025-06-27 at 07 55 07](https://github.com/user-attachments/assets/cc4de201-9541-4d7b-a4d9-86fea33ee9fc)

> ![Request 2](../image_16.png)

Author: ![Screenshot 2025-06-27 at 07 51 29](https://github.com/user-attachments/assets/9a9a242f-971e-4a0b-84f6-1927b4dbe69a)

> ![Request 2](../image_16.png)

Reviewer 4: I've pulled the latest branch but am seeing this still as a png not a json code snippit, will address with my changes

## Multi-part Upload  

After creating a placeholder file through the **Create File (local upload)** endpoint, you will have one or more upload urls. These upload urls are required to make upload requests. It may be useful to compose a shell script that splits up the source file into chunks and issues the same number of subsequent requests.

> **NOTE:** These are important details to keep in mind when sending the subsequent upload requests.

> * The HTTP request method must be `PUT`.
> * The `x-amz-acl` header must be included and be set to private.
> * The `Content-Type` header must match the `media_type` specified in the original **Create File (local upload)** request. This is true even when uploading the file as separate parts.

> In the sample Python script below, we're passing in one upload url in the `upload_urls` parameter, though as noted above, you may have more than one provided to you based on the file size set in the request body for **Create File (local upload).**

Reviewer: `In the sample Python script below, we're passing in multiple upload urls in the `upload_urls` parameter.`

> In the sample Python script below, we're passing in one upload url in the `upload_urls` parameter, though as noted above, you may have more than one provided to you based on the file size set in the request body for **Create File (local upload).**

Reviewer 4: fixed in latest commit

``` python
import requests
import math
from typing import List
from tqdm import tqdm  # For progress bar

> 

Reviewer: May be better here to indicate `chunk_size` is optional and defaults to `none`. @Sahedeva what do you think? 
`def upload_file_in_chunks(file_path: str, upload_urls: list[str], chunk_size: int | None = None) -> bool:`

                    
                    # Read chunk from file
                    f.seek(start_byte)
                    chunk = f.read(end_byte - start_byte)
                    
                    # Upload chunk
                    response = requests.put(
                        url,
                        data=chunk,
                        headers={
                            'Content-Length': str(len(chunk)),
                            'Content-Type': 'image/jpeg',
                            'X-Amz-Acl': 'private'
                            }
                    )
                    
                    if response.status_code != 200:
                        print(f"Failed to upload chunk {i+1}. Status code: {response.status_code}")
                        return False
                    
                    pbar.update(1)
        
        return True
    
    except Exception as e:
        print(f"Error during upload: {str(e)}")
        return False

# Example usage
if __name__ == "__main__":
    # Replace these with your actual values
    file_path =  "/Users/MyComputer/local_upload/sample.jpg"  # Path to your file

>     upload_urls = ["https://frameio-uploads-development.s3-accelerate.amazonaws.com/uploads/10e35b09-7e07-43a1-bd43-897f252ab627/original.png?..."]

Reviewer 3: while this is using the url that was returned in the response, it might be a good example to include how to capture that JSON response and then add that as a variable here to show the whole experience from Auth to Upload.

Reviewer: ```
upload_urls = ["https://frameio-uploads-development.s3-accelerate.amazonaws.com/parts/10e35b09-7e07-43a1-bd43-897f252ab627/part_1?...", "https://frameio-uploads-development.s3-accelerate.amazonaws.com/parts/10e35b09-7e07-43a1-bd43-897f252ab627/part_2?...", "https://frameio-uploads-development.s3-accelerate.amazonaws.com/parts/10e35b09-7e07-43a1-bd43-897f252ab627/part_3?..."]

Author: ![Screenshot 2025-06-27 at 07 56 50](https://github.com/user-attachments/assets/0a662a65-9df3-4dd9-bc33-926cce59303a)

Reviewer 4: resolved in latest commit

        "parent_id": "2e426fe0-f965-4594-8b2b-b4dff1dc00ec",
        "project_id": "7e46e495-4444-4555-8649-bee4d391a997",
        "created_at": "2025-06-26T20:08:06.751313Z",
        "upload_urls": [
            {
                "size": 16881997,
                "url": "https://frameio-uploads-development.s3-accelerate.amazonaws.com/parts/fa18ba7b-b3ee-4dd6-9b31-bd07e554241d/part_1?..."
            },
            {
                "size": 16881997,
                "url": "https://frameio-uploads-development.s3-accelerate.amazonaws.com/parts/fa18ba7b-b3ee-4dd6-9b31-bd07e554241d/part_2?..."
            },
            {
                "size": 16881996,
                "url": "https://frameio-uploads-development.s3-accelerate.amazonaws.com/parts/fa18ba7b-b3ee-4dd6-9b31-bd07e554241d/part_3?..."
            },

        ],
        "view_url": "https://next.frame.io/project/7e46e495-4444-4555-8649-bee4d391a997/view/fa18ba7b-b3ee-4dd6-9b31-bd07e554241d"
    }
}
```

**NOTE:** These are important details to keep in mind when sending the subsequent upload request(s).

> * The HTTP request method must be `PUT`.
> * The `x-amz-acl` header must be included and be set to private.

> > * The `Content-Type` header must match the `media_type` specified in the original **Create File (local upload)** request. This is true even when uploading the file as separate parts.

Reviewer: ```suggestion
> * The `Content-Type` header must match the `media_type` specified in the original **Create File (local upload)** request. This is true even when uploading the file as separate parts. In the example above, the value for `media_type` is `image/jpeg`, so the value for `Content-Type` must also be `image/jpeg`.
```

**Multi-part Uploads**

When a given file results in more than one upload url, it may be useful to compose a shell script that splits up the source file into chunks and issues the same number of subsequent requests.

> In the sample Python script below, we're passing in one upload url in the `upload_urls` parameter, though as noted above, you may have more than one provided to you based on the file size set in the request body for **Create File (local upload).**

Reviewer: The script actually has three upload urls, since I updated the script to match the response body we provided.

> In the sample Python script below, we're passing in one upload url in the `upload_urls` parameter, though as noted above, you may have more than one provided to you based on the file size set in the request body for **Create File (local upload).**

Reviewer 4: ah yes! good catch! I committed the change to include all of the URLs but then forgot to update the copy to reflect!

                        print(f"Response text: {response.text}")
                        print(f"Response headers: {dict(response.headers)}")
                        return False
                    else:
                        print(f"Chunk {i+1} uploaded successfully!")
                    
                    pbar.update(1)
        
        return True
    
    except Exception as e:
        print(f"Error during upload: {str(e)}")
        return False

# Example usage
if __name__ == "__main__":
    # Replace these with your actual values
    file_path =  "/Users/MyComputer/local_upload/sample.jpg"  # Path to your file
    upload_urls = ["https://frameio-uploads-development.s3-accelerate.amazonaws.com/parts/fa18ba7b-b3ee-4dd6-9b31-bd07e554241d/part_1?...", "https://frameio-uploads-development.s3-accelerate.amazonaws.com/parts/fa18ba7b-b3ee-4dd6-9b31-bd07e554241d/part_2?...", "https://frameio-uploads-development.s3-accelerate.amazonaws.com/parts/fa18ba7b-b3ee-4dd6-9b31-bd07e554241d/part_3?..."]
    # Optional: specify chunk size in bytes (e.g., 10MB = 10 * 1024 * 1024)
    content_type = "image/jpeg"

>     chunk_size = 10 * 1024 * 1024  # 10MB

Reviewer 4: going to remove chunk size completely since this script has logic to calculate the chunk size for you.
